pkg/cdi/producer/api.go

```diff
@@ -0,0 +1,37 @@
+/*
+   Copyright © 2024 The CDI Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package producer
+
+import cdi "tags.cncf.io/container-device-interface/specs-go"
+
+// A SpecFormat defines the encoding to use when reading or writing a CDI specification.
+type SpecFormat string
+
+const (
+	// DefaultSpecFormat defines the default encoding used to write CDI specs.
+	DefaultSpecFormat = SpecFormatYAML
+
+	// SpecFormatJSON defines a CDI spec formatted as JSON.
+	SpecFormatJSON = SpecFormat(".json")
+	// SpecFormatYAML defines a CDI spec formatted as YAML.
+	SpecFormatYAML = SpecFormat(".yaml")
+)
+
+// A SpecValidator is used to validate a CDI spec.
+type SpecValidator interface {
+	Validate(*cdi.Spec) error
+}
```

Comment:
I don't want to add unnecessary things to the `specs-go` package. I think we made the call that code to determine the minimum version was close enough to the spec to implement there, but I think an operation on the spec such as validation is not.

Note that due to the nature of go interfaces, this type need not be exported at all -- although it may be useful if it is.

Reply:
Sorry for not being clear with this. I didn't mean moving only SpecValidator interface. I was talking about moving the whole validator package there. The idea was that it doesn't logically belong to producer. Having it here looks a bit confusing to me. 

There is at least one reason not to move it to the spec module though. If it brings a lot of additional external dependencies to the spec package, we should probably avoid doing this. 

@klihub @elezar WDYT?

Reply:
I would rather suggest that we move the version specifics out of `specs-go` implement them as part of the "producer" package. Ideally we would implement this as a submodule as well so that we can properly separate the dependencies.

In the end we will have `specs-go` defining the actual data structures and helpers around this -- including validation. Implemented in other packages / submodules.

Reply:
> Sorry for not being clear with this. I didn't mean moving only SpecValidator interface. I was talking about moving the whole validator package there. The idea was that it doesn't logically belong to producer. Having it here looks a bit confusing to me.


> There is at least one reason not to move it to the spec module though. If it brings a lot of additional external dependencies to the spec package, we should probably avoid doing this.
> 
> @klihub @elezar WDYT?

Yes, I had the same gut feeling. As long as the producer and the consumer were under a single package, having the validator as a subpackage thereof felt like a logical choice. Now if we split them, it does not feel so natural to have it under either... so maybe a `pkg/validator`. This can be adjusted with subsequent PRs, too.

Reply:
I was actually wondering whether we shouldn't make this split at a module level so as to better manage dependencies. Meaning that we'd have:
```
tags.cncf.io/container-device-interface/api/v1/consumer
tags.cncf.io/container-device-interface/api/v1/producer
```
as top-level go submodules.

For the producer specifically, we would start out with:
```
tags.cncf.io/container-device-interface/api/v1/producer/cdi
```
with the types in this package.

We could keep the existing `pkg/cdi` package to provide backward compatibility.

A client that produces specs would then be something like:
```
package main
import (
  "tags.cncf.io/container-device-interface/api/v1/producer/cdi"
)

func main() {
    raw := generateRawCDISpec()
    saver, _ := cdi.New(raw, cdi.WithPermissions(0644))
    err := saver.Save("/var/run/cdi/example.json)
}
```

Reply:
> A client that produces specs would then be something like:
> 
> ```
> package main
> import (
>   "tags.cncf.io/container-device-interface/api/v1/producer/cdi"
> )
> 
> func main() {
>     raw := generateRawCDISpec()
>     saver, _ := cdi.New(raw, cdi.WithPermissions(0644))
>     err := saver.Save("/var/run/cdi/example.json)
> }
> ```

As abstractions go, it feels a bit artificial to have a 1-to-1 association between a producer and a Spec, IOW to essentially make `SpecProducer` a simple Spec wrapper. I would expect a producer to produce one or more Spec files which then typically share the same file format and permissions. So I would make those options, and I'd rather pass the raw spec to be written to the Save or some similar function and not to the constructor itself.

Reply:
That's a fair point. One question in this regard, let's say that the producer is configured to set the minimum version of the spec on save is it OK to modify the spec in-place?

Reply:
> For the producer specifically, we would start out with: tags.cncf.io/container-device-interface/api/v1/producer/cdi

It'd look a bit inconsistent that producer and consumer APIs have a bit different paths. I'd propose to use
 tags.cncf.io/container-device-interface/api/v1/producer and tags.cncf.io/container-device-interface/api/v1/consumer as [you've proposed above](https://github.com/cncf-tags/container-device-interface/pull/233#discussion_r1817723622).